deepdiff/diff.py

```diff
@@ -558,6 +561,53 @@ def _diff_iterable(self, level, parents_ids=frozenset(), _original_type=None):
         else:
             self._diff_iterable_in_order(level, parents_ids, _original_type=_original_type)
 
+    def _compare_in_order(self, level):
+        """
+        Default compare if `iterable_compare_func` is not provided.
+        This will compare in sequence order.
+        """
+
+        return [((i, i), (x, y)) for i, (x, y) in enumerate(
+            zip_longest(
+                level.t1, level.t2, fillvalue=ListItemRemovedOrAdded))]
+
+    def _get_matching_pairs(self, level):
+        """
+        Given a level get matching pairs. This returns list of two tuples in the form:
+        [
+          (t1 index, t2 index), (t1 item, t2 item)
+        ]
+
+        This will compare using the passed in `iterable_compare_func` if available.
+        Default it to compare in order
+        """
+
+        if(self.iterable_compare_func is None):
+            # Match in order if there is no compare function provided
+            return self._compare_in_order(level)
+        try:
+            matches = []
+            y_matched = set()
+            for i, x in enumerate(level.t1):
+                x_found = False
+                for j, y in enumerate(level.t2):
+
+                    if(self.iterable_compare_func(x, y)):
+                        y_matched.add(id(y))
+                        matches.append(((i, j), (x, y)))
+                        x_found = True
+                        break
```

Comment:
What do you want if matches could exist between an item from the iterable 1 to multiple items in the iterable 2? Maybe we should keep track of hashes of what we have already compared and not report it another change or move?

For example this test I wrote fails at the end:

```python
    def test_compare_func2(self):
        t1 = [{'id': 1, 'val': 1}, {'id': 2, 'val': 2}, {'id': 1, 'val': 3}, {'id': 3, 'val': 3}]
        t2 = [{'id': 3, 'val': 3}, {'id': 2, 'val': 2}, {'id': 1, 'val': 3}]
        ddiff = DeepDiff(t1, t2, iterable_compare_func=compare_func)
        expected = {'values_changed': {"root[0]['val']": {'new_value': 3, 'old_value': 1}}, 'iterable_item_moved': {'root[0]': {'new_path': 'root[2]', 'new_value': {'id': 1, 'val': 3}}, 'root[3]': {'new_path': 'root[0]', 'new_value': {'id': 3, 'val': 3}}}}
        assert expected == ddiff
        delta = Delta(ddiff)
        recreated_t2 = t1 + delta
        assert t2 == recreated_t2  # fails
        # The final recreated_t2 will be:
        # [{'id': 3, 'val': 3}, {'id': 2, 'val': 2}, {'id': 1, 'val': 3}, {'id': 1, 'val': 3}]
```

Reply:
Good call. I made a change allowing for a one-to-one relationship between two sets of iterables. If the objects are repeated it will mark them as added or removed.

There is one case where if you move two objects with the same ID it will mark them as values changed rather than iterable item moved. This is consistent with the current behavior. (see test case `test_compare_func_swap`)